README.md
# [MarkDown Box Drawing](https://marketplace.visualstudio.com/items?itemName=TomasHubelbauer.vscode-box-drawing)
![Installs](https://vsmarketplacebadge.apphb.com/installs-short/TomasHubelbauer.vscode-box-drawing.svg)
![Build](https://api.travis-ci.org/TomasHubelbauer/vscode-box-drawing.svg?branch=master)

> A VS Code extension for simplifying box drawing with Unicode and ASCII.

Unfortunately, due to VS Code API limitations, direct mouse drawing mode cannot be used and instead selections have to be utilized.

To alleviate the pain of using this mode, we'll soon bring in a command which pre-spaces a page of the document to make making selections convenient.

![Screenshot](screenshot.gif)

## Features

See the [feature showcase](demo\README.md)

## Release Notes

See the [change log](CHANGELOG.md).

## Contributing

See the [contributor's guide](CONTRIBUTING.md).


CHANGELOG.md
# Change Log

## [Unreleased]

- Respect the `boxDrawing.style` configuration setting when drawing boxes

## `3.0.1` (2018-05-13)

- Switch to new box drawing algorithm
- Improve tests
- Use Unicode middle dot instead of a dash when drawing arrows

## `3.0.0` (2018-05-09)

- Simplify the code to use an editor command
- Contribute a context menu command *Draw a box enclosing selection*
- Contribute a new context menu command *Insert a drawing canvas* which inserts a 120 characters by 30 lines canvas of spaces
- Bind `Ctrl/Cmd+D+B` to drawing a box and `Ctrl/Cmd+I+C` to inserting the canvas
- Implement rudimentary arrow drawing between selection points using the new *Draw an arrow* command (`Ctrl/Cmd+D+A`)

## `2.0.0` (2018-04-11)

- Drop the *VSCode* prefix.

## `1.0.0` (2018-04-04)

- Introduce the *Draw a box* command which uses Unicode box drawing characters


src/test/extension.test.ts
//
// Note: This example test is leveraging the Mocha test framework.
// Please refer to their documentation on https://mochajs.org/ for help.
//

import * as assert from 'assert';
import { workspace, window, Selection, Position, commands } from 'vscode';

suite("Extension Tests", function () {
        test("Box 1", async function () {
                const content = `
a b c d e f g
h i j k l
m n o p q r s t
u v w x y z
`;
                const expected = `
a b c d e f g
h ┏━━┓k l
m ┃ o┃p q r s t
u ┗━━┛x y z
`;
                const textDocument = await workspace.openTextDocument({ language: 'markdown', content });
                const textEditor = await window.showTextDocument(textDocument);
                textEditor.selection = new Selection(new Position(2, 2), new Position(4, 6));
                await commands.executeCommand('extension.drawBox');
                await new Promise(resolve => setTimeout(resolve, 2)); // Prevent flaky test.
                const actual = textDocument.getText();
                assert.equal(actual, expected);
        });

        test("Box 2", async function () {
                const content = `
abcdefghijklmnopqrstuvwxyz
abcdefghijklmnopqrstuvwxyz
abcde
abcdefghijklmnopqrstuvwxyz
abcde
abcdefghijklmnopqrstuvwxyz
abcdefghijklmnopqrstuvwxyz
`;
                const expected = `
ab┏━━━━━━━━━━━━━━━━━┓vwxyz
ab┃defghijklmnopqrst┃vwxyz
ab┃de               ┃
ab┃defghijklmnopqrst┃vwxyz
ab┃de               ┃
ab┗━━━━━━━━━━━━━━━━━┛vwxyz
abcdefghijklmnopqrstuvwxyz
`;
                const textDocument = await workspace.openTextDocument({ language: 'markdown', content });
                const textEditor = await window.showTextDocument(textDocument);
                textEditor.selection = new Selection(new Position(1, 2), new Position(6, 21));
                await commands.executeCommand('extension.drawBox');
                await new Promise(resolve => setTimeout(resolve, 2)); // Prevent flaky test.
                const actual = textDocument.getText();
                assert.equal(actual, expected);
        });

        test("Box 3", async function () {
                const content = `
abcde
abcdefghijklmnopqrstuvwxyz
abcdefghijklmnopqrstuvwxyz
abcdefghijklmnopqrstuvwxyz
abcdefghijklmnopqrstuvwxyz
abcde
end
`;
                const expected = `
ab┏━━━━━━━━━━━━━━━━━┓
ab┃defghijklmnopqrst┃vwxyz
ab┃defghijklmnopqrst┃vwxyz
ab┃defghijklmnopqrst┃vwxyz
ab┃defghijklmnopqrst┃vwxyz
ab┗━━━━━━━━━━━━━━━━━┛
end
`;
                const textDocument = await workspace.openTextDocument({ language: 'markdown', content });
                const textEditor = await window.showTextDocument(textDocument);
                textEditor.selection = new Selection(new Position(1, 2), new Position(6, 21));
                await commands.executeCommand('extension.drawBox');
                await new Promise(resolve => setTimeout(resolve, 2)); // Prevent flaky test.
                const actual = textDocument.getText();
                assert.equal(actual, expected);
        });

        test("Box 4", async function () {
                const content = `
abcdefghi
1
2bcdefghi
3bcdefghi
4
abcdefghi
`;
                const expected = `
a┏━━━┓ghi
1┃   ┃
2┃cde┃ghi
3┃cde┃ghi
4┃   ┃
a┗━━━┛ghi
`;
                const textDocument = await workspace.openTextDocument({ language: 'markdown', content });
                const textEditor = await window.showTextDocument(textDocument);
                textEditor.selection = new Selection(new Position(1, 1), new Position(6, 6));
                await commands.executeCommand('extension.drawBox');
                await new Promise(resolve => setTimeout(resolve, 2)); // Prevent flaky test.
                const actual = textDocument.getText();
                assert.equal(actual, expected);
        });

        test('Arrow 1', async function () {
                const content = `
abcdefghijklmnopqrstuvwxyz
abcdefghijklmnopqrstuvwxyz
abcdefghijklmnopqrstuvwxyz
abcdefghijklmnopqrstuvwxyz
abcdefghijklmnopqrstuvwxyz
abcdefghijklmnopqrstuvwxyz
`;
                const expected = `
a··defghijklmnopqrstuvwxyz
abc····hijklmnopqrstuvwxyz
abcdefg····lmnopqrstuvwxyz
abcdefghijk····pqrstuvwxyz
abcdefghijklmno····tuvwxyz
abcdefghijklmnopqrs··vwxyz
`;
                const textDocument = await workspace.openTextDocument({ language: 'markdown', content });
                const textEditor = await window.showTextDocument(textDocument);
                textEditor.selection = new Selection(new Position(1, 1), new Position(6, 20));
                await commands.executeCommand('extension.drawArrow');
                await new Promise(resolve => setTimeout(resolve, 2)); // Prevent flaky test.
                const actual = textDocument.getText();
                assert.equal(actual, expected);
        });
});


algo.js
// node algo.js
const { performance } = require('perf_hooks');

const EndOfLine = {
    LF: 1,
    CRLD: 2
};

const runs = 100;

true && test(EndOfLine.LF,
    `a b c d e f g
h i j k l
m n o p q r s t
u v w x y z`,
    1, 2, 3, 6,
    `a b c d e f g
h ┏━━┓k l
m ┃ o┃p q r s t
u ┗━━┛x y z`);

true && test(EndOfLine.LF,
    `
abcdefghijklmnopqrstuvwxyz
abcdefghijklmnopqrstuvwxyz
abcde
abcdefghijklmnopqrstuvwxyz
abcde
abcdefghijklmnopqrstuvwxyz
abcdefghijklmnopqrstuvwxyz
`,
    1, 2, 6, 21,
    `
ab┏━━━━━━━━━━━━━━━━━┓vwxyz
ab┃defghijklmnopqrst┃vwxyz
ab┃de               ┃
ab┃defghijklmnopqrst┃vwxyz
ab┃de               ┃
ab┗━━━━━━━━━━━━━━━━━┛vwxyz
abcdefghijklmnopqrstuvwxyz
`);

true && test(EndOfLine.LF,
    `
abcde
abcdefghijklmnopqrstuvwxyz
abcdefghijklmnopqrstuvwxyz
abcdefghijklmnopqrstuvwxyz
abcdefghijklmnopqrstuvwxyz
abcde
end
`,
    1, 2, 6, 21,
    `
ab┏━━━━━━━━━━━━━━━━━┓
ab┃defghijklmnopqrst┃vwxyz
ab┃defghijklmnopqrst┃vwxyz
ab┃defghijklmnopqrst┃vwxyz
ab┃defghijklmnopqrst┃vwxyz
ab┗━━━━━━━━━━━━━━━━━┛
end
`);

true && test(EndOfLine.LF,
    `
abcdefghi
1
2bcdefghi
3bcdefghi
4
abcdefghi
`,
    1, 1, 6, 6,
    `
a┏━━━┓ghi
1┃   ┃
2┃cde┃ghi
3┃cde┃ghi
4┃   ┃
a┗━━━┛ghi
`);

function test(eol, sourceText, selectionStartLine, selectionStartCharacter, selectionEndLine, selectionEndCharacter, expectedText) {
    const test = {
        document: {
            eol,
            getText: () => sourceText,
            offsetAt: (position) => {
                let eolText = '';
                switch (eol) {
                    case EndOfLine.LF: eolText = '\n'; break;
                    case EndOfLine.CRLF: eolText = '\r\n'; break;
                    default: throw new Error('Unexpected EOL value.');
                }

                const lines = sourceText.split(new RegExp('\n', 'mg'));
                let offset = 0;
                for (let line = 0; line < position.line; line++) {
                    offset += lines[line].length + eolText.length;
                }

                offset += position.character;
                return offset;
            }
        },
        selection: {
            start: {
                line: selectionStartLine,
                character: selectionStartCharacter
            },
            end: {
                line: selectionEndLine,
                character: selectionEndCharacter
            }
        }
    };

    const times = [];
    for (let run = 0; run < runs; run++) {
        const t = performance.now();
        const { replacementText, boxHeight, boxWidth } = algo(test);
        times.push(performance.now() - t);
    }

    const { replacementText, boxHeight, boxWidth } = algo(test);
    const actualText = sourceText.substring(0, test.document.offsetAt(test.selection.start)) + replacementText + sourceText.substring(test.document.offsetAt(test.selection.end));
    console.log();
    if (expectedText !== actualText) {
        for (let index = 0; index < expectedText.length; index++) {
            if (expectedText[index] !== actualText[index]) {
                console.log('MISMATCH!', `@${index}`, JSON.stringify(expectedText[index]), JSON.stringify(actualText[index]));
                break;
            }
        }

        console.log('---');
        console.log(expectedText);
        console.log('---');
        console.log(actualText);
    } else {
        console.log(`MATCH (${boxWidth}×${boxHeight} box, ${median(times)} ms)`);
    }
}

function median(numbers) {
    // median of [3, 5, 4, 4, 1, 1, 2, 3] = 3
    var median = 0, numsLen = numbers.length;
    numbers.sort();
    if (numsLen % 2 === 0) { // is even
        // average of two middle numbers
        median = (numbers[numsLen / 2 - 1] + numbers[numsLen / 2]) / 2;
    } else { // is odd
        // middle number only
        median = numbers[(numsLen - 1) / 2];
    }
    return median;
}

function algo(editor) {
    const { character: startCharacter, line: startLine } = editor.selection.start;
    const { character: endCharacter, line: endLine } = editor.selection.end;

    const minCharacter = Math.min(startCharacter, endCharacter);
    const maxCharacter = Math.max(startCharacter, endCharacter);
    const boxWidth = maxCharacter - minCharacter;

    const minLine = Math.min(startLine, endLine);
    const maxLine = Math.max(startLine, endLine);
    const boxHeight = maxLine - minLine + 1;

    let eolText = '';
    switch (editor.document.eol) {
        case EndOfLine.LF: eolText = '\n'; break;
        case EndOfLine.CRLF: eolText = '\r\n'; break;
        default: throw new Error('Unexpected EOL value.');
    }

    const selectionText = editor.document.getText().substring(
        editor.document.offsetAt(editor.selection.start),
        editor.document.offsetAt(editor.selection.end));

    let replacementText = '';

    let cursor = selectionText.indexOf(eolText) + eolText.length;

    replacementText += '┏';
    replacementText += '━'.repeat(boxWidth - 2); // Box top side
    replacementText += '┓';
    if (boxWidth > cursor) {
        replacementText += selectionText.substring(cursor - eolText.length, cursor); // Rest of line
    } else {
        replacementText += selectionText.substring(boxWidth, cursor); // Rest of line
    }
    
    for (let line = 1; line < boxHeight; line++) {
        let leftOffset = cursor + minCharacter;
        let rightOffset = leftOffset + boxWidth;
        replacementText += selectionText.substring(cursor, leftOffset); // Before box line part
        cursor = selectionText.indexOf(eolText, cursor) + eolText.length;

        const isLast = line === boxHeight - 1;
        if (!isLast) {
            replacementText += '┃'; // Box left side
            // Box line part
            if (rightOffset > cursor) {
                const boxLine = selectionText.substring(leftOffset + 1, cursor - 1);
                // TODO: Fix this, it's weird, this code must be better generalizeable
                if (boxLine !== eolText) {
                    replacementText += boxLine.padEnd(boxWidth - 2, ' ');
                } else {
                    replacementText += ' '.repeat(boxWidth - 2);
                }

                replacementText += '┃'; // Box right side
                replacementText += selectionText.substring(cursor - eolText.length, cursor); // After box line part
            } else {
                replacementText += selectionText.substring(leftOffset + 1, rightOffset - 1);
                replacementText += '┃'; // Box right side
                replacementText += selectionText.substring(rightOffset, cursor); // After box line part
            }
        } else {
            replacementText += '┗'; // Box left side
            replacementText += '━'.repeat(boxWidth - 2); // Box line part
            replacementText += '┛'; // Box right side
            if (cursor - eolText.length !== -1) {
                replacementText += selectionText.substring(cursor - eolText.length);
            }
        }
    }

    return { replacementText, boxHeight, boxWidth };
}


src/extension.ts
'use strict';
import { ExtensionContext, commands, window, Position, workspace, TextEditorEdit, Range, TextLine } from 'vscode';

export function activate(context: ExtensionContext) {
    context.subscriptions.push(commands.registerTextEditorCommand('extension.insertCanvas', (textEditor, edit) => {
        if (textEditor.document.languageId !== 'markdown') {
            window.showErrorMessage('Must be a MarkDown editor!');
            return;
        }

        // TODO: https://github.com/Microsoft/vscode/issues/14756
        edit.replace(textEditor.selection, `\n${Array(30).fill(' '.repeat(120)).join('\n')}\n`);
    }));

    context.subscriptions.push(commands.registerTextEditorCommand('extension.drawBox', (textEditor, edit) => {
        if (textEditor.document.languageId !== 'markdown' || textEditor.selections.length !== 1) {
            window.showErrorMessage('Must be a MarkDown editor with a sole selection!');
            return;
        }

        const { start: startPosition, end: endPosition } = textEditor.selection;
        const { style } = workspace.getConfiguration('boxDrawing');
        drawBox(startPosition, endPosition, edit, textEditor.document.lineAt, style);
    }));

    context.subscriptions.push(commands.registerTextEditorCommand('extension.drawArrow', (textEditor, edit) => {
        if (textEditor.document.languageId !== 'markdown' || textEditor.selections.length !== 1) {
            window.showErrorMessage('Must be a MarkDown editor with a sole selection!');
            return;
        }

        const { start: startPosition, end: endPosition } = textEditor.selection;
        drawArrow(startPosition, endPosition, edit);
    }));
}

// TODO: Respect the switch for ASCII/Unicode
export function drawBox(startPosition: Position, endPosition: Position, edit: TextEditorEdit, lineAt: (index: number) => TextLine, style: 'ascii' | 'unicode') {
    const { character: startCharacter, line: startLine } = startPosition;
    const { character: endCharacter, line: endLine } = endPosition;

    const nearCharacter = Math.min(startCharacter, endCharacter);
    const farCharacter = Math.max(startCharacter, endCharacter);
    const nearLine = Math.min(startLine, endLine);
    const farLine = Math.max(startLine, endLine);

    for (let index = nearLine; index <= farLine; index++) {
        const line = lineAt(index);

        const nearSymbol = line.lineNumber === nearLine ? '┏' : (line.lineNumber === farLine ? '┗' : '┃');
        const farSymbol = line.lineNumber === nearLine ? '┓' : (line.lineNumber === farLine ? '┛' : '┃');
        if (line.range.end.character < nearCharacter) {
            // Pad the line long enough to reach the box
            edit.insert(line.range.end, ' '.repeat(nearCharacter - line.range.end.character));
            const position = new Position(line.lineNumber, nearCharacter);
            if (line.lineNumber !== nearLine && line.lineNumber !== farLine) {
                edit.insert(position, '┃' + ' '.repeat(farCharacter - nearCharacter - 2) + '┃');
            } else {
                edit.insert(position, nearSymbol + '━'.repeat(farCharacter - nearCharacter - 2) + farSymbol);
            }
        } else if (line.range.end.character < farCharacter) {
            // End line at the box for simple insertion
            const deleteRange = new Range(new Position(line.lineNumber, nearCharacter), line.range.end);
            edit.delete(deleteRange);
            const insertPosition = line.range.end;
            if (line.lineNumber !== nearLine && line.lineNumber !== farLine) {
                const text = line.text.substring(nearCharacter + 1);
                edit.insert(insertPosition, nearSymbol + text + ' '.repeat(farCharacter - nearCharacter - text.length - 2) + farSymbol);
            } else {
                edit.insert(insertPosition, nearSymbol + '━'.repeat(farCharacter - nearCharacter - 2) + farSymbol);
            }
        } else {
            const range = new Range(new Position(line.lineNumber, nearCharacter), new Position(line.lineNumber, farCharacter));
            if (line.lineNumber !== nearLine && line.lineNumber !== farLine) {
                edit.replace(range, nearSymbol + line.text.substring(nearCharacter + 1, farCharacter - 1) + farSymbol);
            } else {
                edit.replace(range, nearSymbol + '━'.repeat(farCharacter - nearCharacter - 2) + farSymbol);
            }
        }
    }
}

export function drawArrow(startPosition: Position, endPosition: Position, edit: TextEditorEdit) {
    let { line: y0, character: x0 } = startPosition;
    let { line: y1, character: x1 } = endPosition;

    // https://stackoverflow.com/a/4672319/2715716
    const dx = Math.abs(x1 - x0);
    const dy = Math.abs(y1 - y0);
    const sx = (x0 < x1) ? 1 : -1;
    const sy = (y0 < y1) ? 1 : -1;
    let err = dx - dy;
    while (true) {
        // TODO: Test edge cases where x0 + 1 might come bite our ass.
        edit.replace(new Range(new Position(y0, x0), new Position(y0, x0 + 1)), '·');

        if ((x0 === x1) && (y0 === y1)) {
            break;
        }

        const e2 = 2 * err;

        if (e2 > -dy) {
            err -= dy;
            x0 += sx;
        }

        if (e2 < dx) {
            err += dx;
            y0 += sy;
        }
    }
}
